File: generator/src/main/java/ch/bailu/gtk/model/ModelList.java
package ch.bailu.gtk.model;

import java.util.ArrayList;
import java.util.List;

public class ModelList<T extends Model> extends ArrayList<T> {

    private final List<Model> unsupported;

    public ModelList() {
        this.unsupported = new ArrayList<>();
    }

    public ModelList(List<Model> unsupported) {
        this.unsupported = unsupported;
    }


    public void addIfSupported(T model) {
        if (model.isSupported()) {
            add(model);
        } else {
            unsupported.add(model);
        }
    }


}

File: generator/src/main/kotlin/ch/bailu/gtk/model/filter/ModelList.kt
package ch.bailu.gtk.model.filter

import ch.bailu.gtk.model.Model

class ModelList<T : Model>(private val unsupported: ArrayList<Model> = ArrayList()) : java.util.ArrayList<T>() {

    fun addIfSupported(model: T) {
        addIfSupported(model, model.isSupported)
    }

    fun addIfSupported(model: T, supported: Boolean) {
        if (supported) {
            add(model)
        } else {
            unsupported.add(model)
        }
    }

    override fun add(element: T): Boolean {
        return if (contains(element)) {
            false
        } else {
            super.add(element)
        }
    }
}
